JDK 17 Removed some obsolete class file format in regards to runtime
visible annotations.

Now, the annotation parser expects a class type signature, instead of a
class, in the constant pool.

See the JDK commit "Remove vestiages of intermediate JSR 175 annotation format"
https://github.com/openjdk/jdk/commit/8758b554a089af98fdc3be4201d61278d5f01af3
--- a/src/org/armedbear/lisp/jvm-class-file.lisp
+++ b/src/org/armedbear/lisp/jvm-class-file.lisp
@@ -1686,9 +1686,17 @@ appropriate reflective APIs.")
   (dolist (ann (annotations-list annotations))
     (finalize-annotation ann class)))
 
+;; SignatureParser.java
+(defun class-name->type-signature (class-name)
+  (concatenate 'string
+               '(#\L)
+               (substitute #\/ #\. class-name)
+               '(#\;)))
+
 (defun finalize-annotation (ann class)
   (setf (annotation-type ann)
-        (pool-add-class (class-file-constants class) (annotation-type ann)))
+        (pool-add-utf8 (class-file-constants class)
+                       (class-name->type-signature (annotation-type ann))))
   (dolist (elem (annotation-elements ann))
     (finalize-annotation-element elem class)))
 
diff --git a/test/lisp/abcl/runtime-class.lisp b/test/lisp/abcl/runtime-class.lisp
index 856fb781..45096609 100644
--- a/test/lisp/abcl/runtime-class.lisp
+++ b/test/lisp/abcl/runtime-class.lisp
@@ -73,3 +73,14 @@
      0
      20)
   "#<FooList [Foo, Foo,")
+
+
+;; class annotations
+(deftest runtime-class.class-annotations
+  (let* ((class (java:jnew-runtime-class
+                "Foo"
+                :annotations '("java.lang.Deprecated")))
+         (annotations (java:jcall "getAnnotations" class)))
+    (assert (java:jinstance-of-p (aref annotations 0) "java.lang.Deprecated"))
+    t)
+  t)
